docs: Update API docs

Dan-Heath opened the pull request:
This PR adds new content to describe the API pagination feature. To make room for the topic, I broke up the existing API topic into 2 new topics.

- [Overview](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/api-clients/api) - contains general information about the API. I added a new section for **[Manage system resources](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/api-clients/api#manage-system-resources)** to help users navigate the new content.
- [Rate limiting](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/api-clients/api/rate-limiting) - contains information about the recent rate limiting feature, this content is unchanged.

The new content is mostly contained in [API list pagination](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/api-clients/api/pagination) page.

I also updated the following CLI command pages to include the new `list` parameters:

- [accounts list](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/commands/accounts/list)
- [auth-methods list](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/commands/auth-methods/list)
- [auth-tokens list](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/commands/auth-tokens/list)
- [credential-libraries list](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/commands/credential-libraries/list)
- [credential-stores list](https://boundary-23r50lp6n-hashicorp.vercel.app/boundary/docs/commands/credential-stores/list)
- [credentials list](https://boundary-23r50lp6n-hashicorp.vercel.app/boundary/docs/commands/credentials/list)
- [groups list](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/commands/groups/list)
- [host-catalogs list](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/commands/host-catalogs/list)
- [host-sets list](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/commands/host-sets/list)
- [hosts list](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/commands/hosts/list)
- [managed-groups list](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/commands/managed-groups/list)
- [roles list](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/commands/roles/list)
- [scopes list](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/commands/scopes/list)
- [session-recordings list](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/commands/session-recordings/list)
- [sessions list](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/commands/sessions/list)
- [storage-buckets list](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/commands/storage-buckets/list)
- [targets list](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/commands/targets/list)
- [users list](https://boundary-e2848jx0r-hashicorp.vercel.app/boundary/docs/commands/users/list)

johanbrandhorst commented on website/content/docs/api-clients/api/pagination.mdx:
```diff
@@ -0,0 +1,129 @@
+
+Boundary also includes the top-level response `deleted_ids` as part of the refresh operation.
+The `deleted_ids` response contains the IDs of any resources that were deleted since the initial `list` operation.
+Deleted resources are included in the results of the list for 30 days after they are deleted.
+Resources that have been deleted include the suffix `_deleted`.
```
There appears to be some confusion here. The deleted resources returned are always relative to the start of the previous "pagination phase". E.g. for the first "refresh" phase it will, on the first page, include the IDs of any resources deleted since the start of the pagination. To illustrate:

```
1. Start paginating
2. A resource is deleted
3. Finish paginating
4. Start a "refresh phase". This response will include the ID of the resource deleted in 2.
5. Another resource is deleted.
6. Finish paginating the refresh phase.
7. Start another refresh phase. This response will include the ID of the resource deleted in 5.
```

Does that make sense?

Dan-Heath replied:
Makes sense, thank you! I changed this section to read:

> Boundary also includes the top-level response `removed_ids` as part of the refresh operation.
> The `removed_ids` response contains the IDs of any resources that were deleted since you began paginating the results of the previous `list` operation.
> Any resources that have been deleted are included on the first page of results.

johanbrandhorst commented on website/content/docs/api-clients/api/pagination.mdx:
```diff
@@ -0,0 +1,129 @@
+## Refresh the local cache
+
+When the `response_type` reads `complete`, you have viewed all the results that were available when you ran the initial `list` operation.
+You can now refresh the cached list of results to return any resources that were updated since you started viewing the list responses.
```
It's also worth adding here that these results are ordered by `update_time` descending (most recently updated first).

johanbrandhorst commented on website/content/docs/api-clients/api/pagination.mdx:
```diff
@@ -0,0 +1,129 @@
+The value is `delta` until the last page is returned, at which point it will be `complete`.
+- `list_token` - An opaque token which is included in subsequent listing requests after the initial one, so that the client can continue to paginate the results.
+- `sort_by` - The column name that the returned items should be sorted on.
+Each list endpoint has one defined column which defines the sorting for all returned items.
```
I don't know that we need this sentence. It'll always be either `created_time` or `updated_time`, but I don't think we need to say that in the docs.

johanbrandhorst commented on website/content/docs/api-clients/api/pagination.mdx:
```diff
@@ -0,0 +1,129 @@
+- `sort_dir` - The direction in which the returned items are sorted.
+Responses can include `desc` for descending, or `asc` for ascending.
+- `items` - The list of items.
+The items represent replacements for any existing items with matching IDs in the client's cache.
```
This is not right, it's just a list of items. I think we can remove this.

johanbrandhorst commented on website/content/docs/api-clients/api/pagination.mdx:
````diff
@@ -0,0 +1,129 @@
+```json
+{
+  "response_type": "<delta|complete>",
+  "pagination_token": "<opaque-token>",
````
```suggestion
  "list_token": "<opaque-token>",
```

johanbrandhorst commented on website/content/docs/api-clients/api/pagination.mdx:
```diff
@@ -0,0 +1,129 @@
+
+## List tokens
+
+List tokens contain general metadata associated with the ongoing pagination and either a pagination or refresh token, depending on the phase of the pagination.
```
This contains internal implementation details that I don't think we need to mention
```suggestion
List tokens contain general metadata associated with the ongoing pagination.
```

johanbrandhorst commented on website/content/docs/api-clients/api/pagination.mdx:
```diff
@@ -0,0 +1,129 @@
+List tokens contain general metadata associated with the ongoing pagination and either a pagination or refresh token, depending on the phase of the pagination.
+If you compose an invalid token, you receive an invalid token error.
+
+When you make a `list` request without a `list_token`, Boundary builds a page of items that contains all the resources currently available, subject to the page size, with the following attributes:
+
+- Must contain a `response_type` equal to `delta` or `complete`
+- Must contain a `sort_by`
+- Must contain a `sort_dir`
+- May contain a `list_token`
+- May contain `items`
+- May contain `est_item_count`
+
+When you make a `list` request using a valid `list_token`, Boundary builds a page of items following from the page returned with the token:
+
+- Must contain a `response_type` equal to `delta` or `complete`
+- Must contain a `sort_by`
+- Must contain a `sort_dir`
+- Must contain a `list_token`
+- May contain `items`
+- May contain `est_item_count`
```
I think we can drop this, it doesn't add that much new.

johanbrandhorst commented on website/content/docs/api-clients/api/pagination.mdx:
```diff
@@ -0,0 +1,129 @@
+
+## Constraints and limitations
+
+In scenarios where updates are so frequent that the results may change while the refresh operation runs, those updates may not appear at all in the refreshed list.
+If you make frequent updates to a resource, or if you are updating data while you run a `list` operation, HashiCorp recommends that you generate the full list of results instead of performing a refresh.
+
+Workers are not included in refresh operations because the updates they send to the controller are too frequent to be captured in a refresh.
+To update a list of worker resources, you must regenerate the full list of results.
```
This is great. Can we also add a notice that there may be duplicate items returned in the "refresh phase" pages, and that clients need to be able to handle these. This is a fundamental tradeoff when working across multiple database transactions, with the alternative being that sometimes items are missed from the pagination.

Dan-Heath replied:
Sure! I added this:

>When you refresh the list of results, you may notice duplicate items in the list.
This is a fundamental tradeoff when you search across multiple database transactions.

Should we provide any additional follow up advice here? Or is it just something for them to be aware of?

johanbrandhorst replied:
I think this is fine, I just want it to be clear that it is the responsibility of the user to handle duplicates in the responses.

johanbrandhorst reviewed (changes requested):
This is a great start, thank you! I added some comments

Dan-Heath commented on website/content/docs/api-clients/api/pagination.mdx:
```diff
@@ -0,0 +1,129 @@
+
+The results of the `list` operation are cached locally to help protect your system resources.
+Maintaining a local cache of results for a frequently used `list` operation can also help to reduce server cost.
+You can refresh the local results of your initial `list` operation as many times as you want for up to 30 days.
```
This makes sense, thanks! We are definitely planning a standalone page for the client cache demon, and I will be starting work on it soon. I cut this section, and we can add a link to the new client cache demon topic, if it makes sense to later.

Dan-Heath commented:
> This is a great start, thank you! I added some comments

Thank you for the review and feedback, @johanbrandhorst ! All makes sense, and I think I updated everything accordingly. I also added the new `list` parameters to each of the individual resource `list` documentation pages. I think this PR is ready for review.

johanbrandhorst commented on website/content/docs/api-clients/api/index.mdx:
````diff
@@ -82,196 +82,13 @@ The following method conventions are used within Boundary's API:
-}
-```
\ No newline at end of file
+When you search for Boundary resources using the `list` operation, you could potentially receive a very large number of results.
+Boundary uses an API pagination feature to enable you to search and filter large lists of search results without overwhelming your system resources.
+The results of the operation are cached locally, to help reduce server cost and further protect your system resources.
````
Again, not sure this will cause confusion. The automatic caching is a feature of the CLI. The Desktop client also uses it, but it's not an automatic feature of the API itself. Maybe this should just be

```suggestion
When you search for Boundary resources using the `list` operation, you could potentially receive a very large number of results.
Boundary uses an API pagination feature to enable you to search and filter large lists of search results without overwhelming your system resources.
```

Dan-Heath replied:
> Is there already a different area where search and the caching is discussed?

Hi @talanknight , I started the work for search and caching on another branch, although I'm wondering if I should combine them now since it's all related. The draft PR is #4244 (it still needs work).

johanbrandhorst commented on website/content/docs/commands/workers/list.mdx:
```diff
@@ -40,5 +40,11 @@ The default value is `false`.
 - `scope-id=<string>` - The scope from which to list the workers.
 The default value is `global`.
 You can also specify this value using the **BOUNDARY_SCOPE_ID** environment variable.
+- `list_token` (optional) - An opaque token that is returned from the previous `list` response.
+If you do not supply a value, pagination starts from the beginning.
+- `page_size` (optional) - An unsigned integer that indicates the number of items that should be included on the page of search results.
+If you do not provide a value, or if you provide a value of `0`, Boundary uses the default page size of 1000 items.
+If you configure it, the `page_size` value overrides the default page size.
+Controller administrators can also configure a `max_page_size` option that constrains the maximum page size a user can request.
```
The workers list endpoint doesn't support the new pagination workflow, so we can remove these two

Dan-Heath replied:
Good catch, thank you!

johanbrandhorst reviewed (changes requested):
Thanks just two minor notes!
